tutorials/haskell/generics/tutorial.md

```diff
@@ -0,0 +1,584 @@
+representations we've just discussed.
+
+Let's see what these type classes have:
+
+```haskell
+class Generic a where
+  type Rep a :: * -> *
+  from :: a -> Rep a p
+  to   :: Rep a p -> x
+
+class Generic1 f where
+  type Rep1 f :: * -> *
+  from1 :: f p -> Rep1 f p
+  to1   :: Rep1 f p -> f p
+```
+
+`from` and `from1` map values of data types to their generic
+representations. `Rep` and `Rep1` are associated type functions (the feature
+is a part of `TypeFamilies` GHC extension) that take type of data we want to
+manipulate and yield type of its representation. Of course, if we want to
+derive `Functor` instances, we need a way to go back from representation to
+actual value of target data type. This is done via `to` and `to1`. The good
+thing about this, of course, that GHC can derive `Generic` and `Generic1`
+for us automatically with `DeriveGeneric` language extension enabled.
+
+We can open GHCi now and try infer `Rep` of some type:
+
+```haskell
+λ> :t (undefined :: Rep (List a) p)
+(undefined :: Rep (List a) p)
+  :: D1
+       ('MetaData "List" "GenericsTutorial" "main" 'False)
+       (C1 ('MetaCons "Nil" 'PrefixI 'False) U1
+        :+: C1
+              ('MetaCons "Cons" 'PrefixI 'False)
+              (S1
+                 ('MetaSel
+                    'Nothing 'NoSourceUnpackedness 'NoSourceStrictness 'DecidedLazy)
+                 (Rec0 a)
+               :*: S1
+                     ('MetaSel
+                        'Nothing 'NoSourceUnpackedness 'NoSourceStrictness 'DecidedLazy)
+                     (Rec0 (List a))))
+       p
+```
+
+Ah OK, there is just a little bit more to this representation thing…
+
+## Metadata wrappers
+
+Representation also has associated metadata. It may look messy and difficult
+to read, but I'll explain you the logic behind it in a moment, it should
+clarify why it's done like this.
+
+First of all, metadata should not get in our way if we don't want to deal
+with it. Thus all metadata is attached using the same simple wrapper:
+
+```haskell
+newtype M1 i c f p = M1 { unM1 :: f p } -- ‘f p’ is what lives inside, U1 for example
+--         ^ ^
+--         | |
+--         | +---- compiler-generated data type that allows to get meta information
+--         |
+--  type-level tag, see below
+```
+
+The `i` type-level tag can be one of the three:
+
+* `D` for data type metadata `type D1 = M1 D`
+* `C` for constructor metadata `type C1 = M1 C`
+* `S` for record selector metadata `type S1 = M1 S`
+
+It should be understandable why metadata is attached this way. If we want
+to, we can ignore metadata easily, since it's just a wrapper:
+
+```haskell
+f (M1 x) = f x
+```
+
+If we want to look at a particular type of metadata we can specify the `i`
+type-level tag, or we can leave it unspecified to deal with all metadata at
+once.
+
+The `c` type is auto-generated by the compiler and encodes metadata on type
+level. Why not to store the metadata on value level, in `M1` constructor?
```

Comment:
Remove "to" from "Why not to store the metadata on value level, in `M1` constructor?" leaving "Why not store the metadata on value level, in `M1` constructor?"